docs/DevGuide/CompilerLinkerErrors.md

```diff
@@ -0,0 +1,99 @@
+
+# Linker Errors {#understanding_linker_errors}
+
+Possibly the most difficult part of fixing linking errors is understanding what
+they are trying to tell you. Let's take the following example
+```
+error: undefined reference to 'Tensor<DataVector,
+brigand::list<brigand::integral_constant<int, 1> >,
+brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial,
+(IndexType)0> > >
+random_unit_normal<DataVector>(gsl::not_null<std::mersenne_twister_engine<
+unsigned long, 32ul, 624ul, 397ul, 31ul, 2567483615ul, 11ul, 4294967295ul, 7ul,
+2636928640ul, 15ul, 4022730752ul, 18ul, 1812433253ul>*>, Tensor<DataVector,
+brigand::list<brigand::integral_constant<int, 1>,
+brigand::integral_constant<int, 1> >,
+brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial,
+(IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial,
+(IndexType)0> > > const&)' in
+lib/libTest_GeneralizedHarmonic.a(Test_UpwindFlux.cpp.o):
+Test_UpwindFlux.cpp:function (anonymous namespace)::test_upwind_flux_random()
+```
+We can start by splitting out information about different parts of the
+error. First,
+```
+error: undefined reference to
+```
+tells us that we forgot to add a link dependency for a library or
+executable. The next relevant part of information is which library (executable)
+and file the missing function/class was in. This is at the end of the error
+message (where in the error message can depend on your linker, unfortunately):
+```
+lib/libTest_GeneralizedHarmonic.a(Test_UpwindFlux.cpp.o):
+Test_UpwindFlux.cpp:function (anonymous namespace)::test_upwind_flux_random()
+```
+What this means is that the missing link dependency is used in the library
+`Test_GeneralizedHarmonic`, the file `Test_UpwindFlux.cpp`, and the function
+`test_upwind_flux_random()`.
+
+We have now determined what the linker error is (a missing link dependency), and
+in which library, file, and function the missing link dependency is used. We now
+need to understand what the missing link dependency is. Since SpECTRE uses a lot
+of templates, the missing missing reference (link dependency) can be quite
+long. In this case it is:
+```
+'Tensor<DataVector,
+brigand::list<brigand::integral_constant<int, 1> >,
+brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial,
+(IndexType)0> > >
+random_unit_normal<DataVector>(gsl::not_null<std::mersenne_twister_engine<
+unsigned long, 32ul, 624ul, 397ul, 31ul, 2567483615ul, 11ul, 4294967295ul, 7ul,
+2636928640ul, 15ul, 4022730752ul, 18ul, 1812433253ul>*>, Tensor<DataVector,
+brigand::list<brigand::integral_constant<int, 1>,
+brigand::integral_constant<int, 1> >,
+brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial,
+(IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial,
+(IndexType)0> > > const&)'
+```
+In order to make this easier to read, it is recommended to run the code through
+ClangFormat, which gives
+\code{.cpp}
```

Comment:
Missing \startcode or something...code not in code block

Reply:
Sorry I don't know what you mean. Could you expand on what you're saying please?

Reply:
Looks like it compiles in the spectre doxygen, but github doesn't know how to render the block.

Reply:
Ah, yes, this is only valid Doxygen code... Sorry!